graph/preprocessor.go

```diff
@@ -0,0 +1,291 @@
+	for i := len(alias.AliasSrc) - 1; i >= 0; i-- {
+		aliasURI := alias.AliasSrc[i]
+		if util.IsURL(aliasURI) {
+			if err := addAliasFromRemote(alias, aliasURI); err != nil {
+				return err
+			}
+		} else {
+			if err := addAliasFromFile(alias, aliasURI); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Fetches and parses out remote alias files and adds their content
+// to the passed in Alias. Note alias definitions already in alias
+// will not be overwritten.
+func addAliasFromRemote(alias *Alias, url string) error {
+	remoteClient := &http.Client{
+		Timeout: time.Second * 2, // Maximum of 2 secs
+	}
+
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return err
+	}
+
+	res, getErr := remoteClient.Do(req)
+	if getErr != nil {
+		return getErr
+	}
+
+	if res.StatusCode > 299 {
+		httpErr, err := ioutil.ReadAll(res.Body)
+		if err != nil {
+			return err
+		}
+		return errors.New(string(httpErr))
+
+	}
+
+	defer res.Body.Close()
+
+	data, readErr := ioutil.ReadAll(res.Body)
+	if readErr != nil {
+		return readErr
+	}
+
+	return readAliasFromBytes(data, alias)
+}
+
+// Parses out local alias files and adds their content to the passed in
+// Alias. Note alias definitions already in alias will not be
+// overwritten.
+func addAliasFromFile(alias *Alias, fileURI string) error {
+	data, fileReadingError := ioutil.ReadFile(fileURI)
+	if fileReadingError != nil {
+		return fileReadingError
+	}
+	return readAliasFromBytes(data, alias)
+}
+
+// Parses out alias definitions from a given bytes array and appends
+// them to the Alias. Note alias definitions already in alias will
+// not be overwritten even if present in the array.
+func readAliasFromBytes(data []byte, alias *Alias) error {
+	aliasMap := &map[string]string{}
+
+	if err := yaml.Unmarshal(data, aliasMap); err != nil {
+		return err
+	}
+
+	for key, value := range *aliasMap {
+		if _, ok := alias.AliasMap[key]; !ok {
+			alias.AliasMap[key] = value
+		}
+	}
+	return nil
+}
+
+// preprocessString handles the preprocessing (string replacement and resolution)
+// of all aliases in an input yaml (passed in as a string). The resolved aliases are
+// defined in the input alias file.
+func preprocessString(alias *Alias, str string) (string, bool, error) {
+	// alias.loadGlobalDefinitions TODO?
+
+	// Load Remote/Local alias definitions
+	if externalDefinitionErr := alias.loadExternalAlias(); externalDefinitionErr != nil {
+		return "", false, externalDefinitionErr
+	}
+	// Validate alias definitions
+	if improperFormatErr := alias.resolveMapAndValidate(); improperFormatErr != nil {
+		return "", false, improperFormatErr
+	}
+	var out strings.Builder
+	var command strings.Builder
+	ongoingCmd := false
+	changed := false
+
+	// Search and replace all strings with the directive
+	for _, char := range str {
+		if ongoingCmd {
+			if char == alias.directive && command.Len() == 0 { // Escape Character Triggered
```

Comment:
Can you point me to the code which takes care of 
```
- cmd: bash ...
```
i.e. the case without directive?

Reply:
check the func processSteps(alias *Alias, task *Task). Its called at the end of newTaskFromBytes